File: python/test/t_OrthogonalBasis_std.py
#! /usr/bin/env python

import openturns as ot
import openturns.testing as ott

ot.TESTPREAMBLE()

# Polynomial factories
factoryCollection = [
    ot.LaguerreFactory(2.5),
    ot.LegendreFactory(),
    ot.HermiteFactory(),
]
dim = len(factoryCollection)
basisFactory = ot.OrthogonalProductPolynomialFactory(factoryCollection)
print("basisFactory=")
print(basisFactory)
print(basisFactory.__repr_markdown__())
print(basisFactory._repr_html_())

basis = ot.OrthogonalBasis(basisFactory)
print("basis=")
print(basis)

x = [0.5] * dim
enum = basis.getEnumerateFunction()

# Test build by index and build by multi-index produce the same result
for i in range(10):
    f_by_index = basis.build(i)
    indices = enum(i)
    f_by_multiindex = basis.build(indices)
    val_by_index = f_by_index(x)
    val_by_multiindex = f_by_multiindex(x)
    print("i=", i, "f(X)=", val_by_index)
    ott.assert_almost_equal(val_by_index, val_by_multiindex)

# Other factories
factoryCollection2 = [
    ot.OrthogonalUniVariatePolynomialFunctionFactory(ot.LaguerreFactory(2.5)),
    ot.HaarWaveletFactory(),
    ot.FourierSeriesFactory(),
]
dim2 = len(factoryCollection2)
basisFactory2 = ot.OrthogonalProductFunctionFactory(factoryCollection2)
basis2 = ot.OrthogonalBasis(basisFactory2)
print("basis=", basis2)
x2 = [0.5] * dim2
enum2 = basis2.getEnumerateFunction()
for i in range(10):
    f = basis2.build(i)
    print("i=", i, "f(X)=", f(x2))

# Polynomial factories using a collection of distributions
distributionCollection = [
    ot.Normal(),
    ot.TruncatedDistribution(ot.Normal(2.0, 1.5), 1.0, 4.0),
    ot.Uniform(),
]
basisFactory3 = ot.OrthogonalProductPolynomialFactory(distributionCollection)
print("basisFactory=")
print(basisFactory3)
print(basisFactory3.__repr_markdown__())
print(basisFactory3._repr_html_())

File: lib/test/t_OrthogonalBasis_std.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of class OrthogonalBasis for standard methods
 *
 *  Copyright 2005-2026 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

int main(int, char *[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);

  try
  {
    UnsignedInteger dim = 3;
    OrthogonalProductPolynomialFactory::PolynomialFamilyCollection polynomCollection(dim);
    polynomCollection[0] = LaguerreFactory(3.5);
    polynomCollection[1] = LegendreFactory();
    polynomCollection[2] = HermiteFactory();
    OrthogonalProductPolynomialFactory basisFactory(polynomCollection);
    OrthogonalBasis basis(basisFactory);
    Point point(dim, 0.5);

    // Test build by index and build by multi-index produce the same result
    EnumerateFunction enumFunction(basis.getEnumerateFunction());
    for (UnsignedInteger i = 0; i < 10; ++i)
    {
      Function fByIndex(basis.build(i));
      Indices indices(enumFunction(i));
      Function fByMultiIndex(basis.build(indices));
      Point valByIndex(fByIndex(point));
      Point valByMultiIndex(fByMultiIndex(point));
      fullprint << "i=" << i << " f(point)=" << valByIndex << std::endl;
      assert_almost_equal(valByIndex, valByMultiIndex, 1e-14, 0.0);
    }

    // Using multi-indices explicitly
    for (UnsignedInteger i = 0; i < 10; ++i)
    {
      Indices indices(enumFunction(i));
      Function f(basis.build(indices));
      fullprint << "indices=" << indices << " f(point)=" << f(point) << std::endl;
    }
  }
  catch (TestFailed & ex)
  {
    std::cerr << ex << std::endl;
    return ExitCode::Error;
  }

  return ExitCode::Success;
}
